File: harness/fuzz/runner/runner.py
from __future__ import annotations

import subprocess
import sys
import time
from dataclasses import dataclass
from pathlib import Path

from harness_shared.support import qemu_mps2_an386_command, terminate_process


@dataclass(frozen=True)
class ProcessResult:
    process_status: str
    timeout: bool
    elapsed_ms: int


def qemu_command(
    *,
    qemu: str,
    elf: Path,
    plugin: Path,
    manifest: Path,
) -> list[str]:
    return qemu_mps2_an386_command(qemu, elf) + [
        "-plugin",
        f"file={plugin},manifest={manifest}",
    ]


def run_qemu_trial(
    *,
    qemu: str,
    elf: Path,
    plugin: Path,
    manifest: Path,
    done: Path,
    timeout: float,
) -> ProcessResult:
    command = qemu_command(qemu=qemu, elf=elf, plugin=plugin, manifest=manifest)
    start = time.monotonic()
    proc = subprocess.Popen(
        command,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.PIPE,
        text=True,
    )

    deadline = start + timeout
    try:
        while time.monotonic() < deadline:
            if done.exists():
                terminate_process(proc)
                return ProcessResult(
                    process_status="completed",
                    timeout=False,
                    elapsed_ms=elapsed_since_ms(start),
                )
            status = proc.poll()
            if status is not None:
                warn_on_stderr(proc, f"qemu exited with status {status}")
                return ProcessResult(
                    process_status=f"exit:{status}",
                    timeout=False,
                    elapsed_ms=elapsed_since_ms(start),
                )
            time.sleep(0.02)

        terminate_process(proc)
        warn_on_stderr(proc, "qemu timed out")
        return ProcessResult(
            process_status="timeout",
            timeout=True,
            elapsed_ms=elapsed_since_ms(start),
        )
    finally:
        if proc.poll() is None:
            terminate_process(proc)


def warn_on_stderr(proc: subprocess.Popen[str], reason: str) -> None:
    if proc.stderr is None:
        return
    try:
        stderr = proc.stderr.read()
    except ValueError:
        return
    if stderr.strip():
        print(f"warning: {reason}; qemu stderr:\n{stderr.rstrip()}", file=sys.stderr)


def elapsed_since_ms(start: float) -> int:
    return int((time.monotonic() - start) * 1000)

File: harness/tui/internal/fuzz/runner.go
package fuzz

import (
	"context"
	"fmt"
	"io"
	"strings"
	"time"

	"github.com/b-kaf/fault-tolerance-comparison/harness/tui/internal/qemu"
	"github.com/b-kaf/fault-tolerance-comparison/harness/tui/internal/target"
)

// ProcessResult mirrors runner.ProcessResult. Stderr is everything the QEMU
// child wrote to stderr, which carries the plugin's @@FT result record.
type ProcessResult struct {
	ProcessStatus string
	Timeout       bool
	ElapsedMS     int64
	Stderr        string
}

const terminateTimeout = 2 * time.Second

// resultSentinel marks a complete @@FT result record on the plugin's stderr;
// its appearance is the completion signal (the plugin keeps QEMU running, so
// the runner terminates the process once it sees this).
const resultSentinel = "@@FT-END"

// RunQemuTrial mirrors runner.run_qemu_trial: launch QEMU with the plugin,
// poll for the done flag (checked before process exit, since QEMU keeps
// running after the plugin writes it), and classify how the process ended.
// Cancelling ctx terminates the trial early with status "cancelled".
//
// oneInsnPerTB appends -accel tcg,one-insn-per-tb=on, required by the insn-skip
// campaign so a mid-TB PC write removes exactly one instruction; it is gated
// per-campaign because it slows emulation and other modes do not need it.
func RunQemuTrial(ctx context.Context, profile target.Profile, elfPath, plugin, manifest string, pluginArgs []string, timeout time.Duration, oneInsnPerTB bool, warnings io.Writer) (ProcessResult, error) {
	pluginSpec := fmt.Sprintf("file=%s,manifest=%s", plugin, manifest)
	if len(pluginArgs) > 0 {
		pluginSpec += "," + strings.Join(pluginArgs, ",")
	}
	argv := qemu.BaseCommand(profile, elfPath)
	// -accel must precede -plugin: on RISC-V, configuring one-insn-per-tb after
	// the plugin loads leaves the plugin's vcpu_init register enumeration empty
	// (no pc handle), breaking insn-skip. ARM tolerates either order.
	if oneInsnPerTB {
		argv = append(argv, "-accel", "tcg,one-insn-per-tb=on")
	}
	argv = append(argv, "-plugin", pluginSpec)

	start := time.Now()
	proc, err := qemu.Start(argv)
	if err != nil {
		return ProcessResult{}, err
	}
	defer func() {
		if !proc.Exited() {
			proc.Terminate(terminateTimeout)
		}
	}()

	deadline := start.Add(timeout)
	for time.Now().Before(deadline) {
		if ctx.Err() != nil {
			proc.Terminate(terminateTimeout)
			return ProcessResult{
				ProcessStatus: "cancelled",
				Timeout:       false,
				ElapsedMS:     time.Since(start).Milliseconds(),
				Stderr:        proc.Stderr(),
			}, ctx.Err()
		}
		if strings.Contains(proc.Stderr(), resultSentinel) {
			proc.Terminate(terminateTimeout)
			return ProcessResult{
				ProcessStatus: "completed",
				Timeout:       false,
				ElapsedMS:     time.Since(start).Milliseconds(),
				Stderr:        proc.Stderr(),
			}, nil
		}
		if proc.Exited() {
			status := proc.ExitCode()
			warnOnStderr(warnings, proc, fmt.Sprintf("qemu exited with status %d", status))
			return ProcessResult{
				ProcessStatus: fmt.Sprintf("exit:%d", status),
				Timeout:       false,
				ElapsedMS:     time.Since(start).Milliseconds(),
				Stderr:        proc.Stderr(),
			}, nil
		}
		time.Sleep(20 * time.Millisecond)
	}

	proc.Terminate(terminateTimeout)
	warnOnStderr(warnings, proc, "qemu timed out")
	return ProcessResult{
		ProcessStatus: "timeout",
		Timeout:       true,
		ElapsedMS:     time.Since(start).Milliseconds(),
		Stderr:        proc.Stderr(),
	}, nil
}

func warnOnStderr(warnings io.Writer, proc *qemu.Process, reason string) {
	if warnings == nil {
		return
	}
	stderr := strings.TrimSpace(proc.Stderr())
	if stderr != "" {
		fmt.Fprintf(warnings, "warning: %s; qemu stderr:\n%s\n", reason,
			strings.TrimRight(proc.Stderr(), "\n"))
	}
}
